### DISCUSSION TITLE
How to put limits in some parameter and get alarm sound when limit crosses.

### DISCUSSION THREAD

**Discussion by Vaibhav2121** [nasa/openmct#4117] [Community Q&A] [answered]
How to put limits in some parameter monitored in OpenMct and get alarm sound when limit crosses.

**Comment by jvigliotta** [comment, marked as answer]
Hi @Vaibhav2121 checkout our SinewaveLimitProvider for an example: https://github.com/nasa/openmct/blob/master/example/generator/SinewaveLimitProvider.js

Once you create a provider you can add it using `openmct.telemetry.addProvider(provider)`

**Reply by Vaibhav2121** [reply]
hi jivigliotta,
sinewaveLimitProvider provides limit for 1 graph. How can we apply different limits to hundreds of parameter.

**Reply by davetsay** [reply]
@Vaibhav2121 , you can define which parameters your limit provider applies to in `supportsLimits`. The example in `SinewaveLimitProvider` applies to a type, `domainObject.type === 'generator'`. You can change that to apply to a different type, a set of named parameters, or whatever else.

**Reply by Vaibhav2121** [reply]
Dear Sir, 
I have hundred parameters of one domainObject.type, how can i apply limits to each individual parameter of that type, because if I am applying limits, same limits is applicable to all object.

Example following is dictionary.json

{
    "name": "Example Spacecraft",
    "key": "sc",
    "measurements": [
        {
            "name": "Fuel",
            "key": "prop.fuel",
            "values": [
                {
                    "key": "value",
                    "name": "Value",
                    "units": "kilograms",
                    "format": "float",
                    "min": 0,
                    "max": 100,
                    "hints": {
                        "range": 1
                    }
                },
                {
                    "key": "utc",
                    "source": "timestamp",
                    "name": "Timestamp",
                    "format": "utc",
                    "hints": {
                        "domain": 1
                    }
                }
            ]
        },
        {
            "name": "Thrusters",
            "key": "prop.thrusters",
            "values": [
                {
                    "key": "value",
                    "name": "Value",
                    "format": "enum",
                    "enumerations": [
                        {
                            "string": "ON",
                            "value": 1
                        },
                        {
                            "string": "OFF",
                            "value": 0
                        }
                    ],
                    "hints": {
                        "range": 1
                    }
                },
                {
                    "key": "utc",
                    "source": "timestamp",
                    "name": "Timestamp",
                    "format": "utc",
                    "hints": {
                        "domain": 1
                    }
                }
            ]
        },
        {
            "name": "Received",
            "key": "comms.recd",
            "values": [
                {
                    "key": "value",
                    "name": "Value",
                    "units": "bytes",
                    "format": "integer",
                    "hints": {
                        "range": 1
                    }
                },
                {
                    "key": "utc",
                    "source": "timestamp",
                    "name": "Timestamp",
                    "format": "utc",
                    "hints": {
                        "domain": 1
                    }
                }
            ]
        },
        {
            "name": "Sent",
            "key": "comms.sent",
            "values": [
                {
                    "key": "value",
                    "name": "Value",
                    "units": "bytes",
                    "format": "integer",
                    "hints": {
                        "range": 1
                    }
                },
                {
                    "key": "utc",
                    "source": "timestamp",
                    "name": "Timestamp",
                    "format": "utc",
                    "hints": {
                        "domain": 1
                    }
                }
            ]
        },
        {
            "name": "Generator Temperature",
            "key": "pwr.temp",
            "values": [
                {
                    "key": "value",
                    "name": "Value",
                    "units": "℃",
                    "format": "float",
                    "hints": {
                        "range": 1
                    }
                },
                {
                    "key": "utc",
                    "source": "timestamp",
                    "name": "Timestamp",
                    "format": "utc",
                    "hints": {
                        "domain": 1
                    }
                }
            ]
        },
        {
            "name": "Generator Current",
            "key": "pwr.c",
            "values": [
                {
                    "key": "value",
                    "name": "Value",
                    "units": "Amps",
                    "format": "float",
                    "hints": {
                        "range": 1
                    }
                },
                {
                    "key": "utc",
                    "source": "timestamp",
                    "name": "Timestamp",
                    "format": "utc",
                    "hints": {
                        "domain": 1
                    }
                }
            ]
        },
        {
            "name": "Generator Voltage",
            "key": "pwr.v",
            "values": [
                {
                    "key": "value",
                    "name": "Value",
                    "units": "Volts",
                    "format": "float",
                    "hints": {
                        "range": 1
                    }
                },
                {
                    "key": "utc",
                    "source": "timestamp",
                    "name": "Timestamp",
                    "format": "utc",
                    "hints": {
                        "domain": 1
                    }
                }
            ]
        }
    ]
}

Here measurements[values].key is "value" or "utc" .  I am applying limits to "value" in sinewavelimitProvider it is becoming common to all parameters.  How to apply limits to individual parameters of same object type.

**Reply by jvigliotta** [reply]
Hey do you have an example of your current implementation? Might help us better understand. Do you want the limit to only relate to that field of the datum?

**Reply by Vaibhav2121** [reply]
I want to apply different limits to each parameter like different limits to Fuel, Generator temprature, Generator current and Generator voltage. If I apply limits on value the same limits applies to all parameter. how to aplly different limits to different parameters.

**Comment by davetsay** [comment]
@Vaibhav2121 , you could use a map to apply different limit evaluations to different parameters. Something like below.

```
YourProvider.prototype.getLimitEvaluator = function (domainObject) {
    // define limits map
    const evaluateFunctions = {
        parameterKey1: (datum, valueMetadata) => {},
        parameterKey2: (datum, valueMetadata) => {},
    };
    
    // use identifier key or identifier namespace/key to get evaluate function from map
    const evaluateFunction = evaluateFunctions[domainObject.identifier.key];
    
    return {
        evaluate: evaluateFunction
    };
}
```
